File: fixtures/lifecycle-infrastructure-failure/sdk-probe.mjs
// Copyright 2026 Tymofii Pidlisnyi. Apache-2.0 license. See LICENSE.
//
// Records what the published npm `agent-passport-system` package does with each
// group in this family, so SDK-RUNS.md's entries are reproducible rather than
// asserted.
//
// This probe reports. It does not judge. Nothing in draft-pidlisnyi-aps-03 asks
// the SDK for any of these six behaviours, so a `not_supported` line here is an
// absence of a named surface, not a defect report and not a conformance result.
//
// One group does have a real surface. `enforceFreshnessPolicy` takes an
// `AttestationFreshness` carrying the source's own `validAt` and `ttl` plus the
// verifier's own `maxStalenessMs`, which is exactly the two-window shape the
// status_artifact group is about. This probe runs every status_artifact vector
// through it under two of the SDK's documented policy modes and prints what
// came back next to what the fixture expects. Disagreement is reported as
// disagreement; the probe never rewrites either side.
//
// Run: node fixtures/lifecycle-infrastructure-failure/sdk-probe.mjs
// Exit 0 always.

import { readFileSync } from 'node:fs'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'

import * as aps from 'agent-passport-system'

const HERE = dirname(fileURLToPath(import.meta.url))
const vectors = JSON.parse(readFileSync(join(HERE, 'vectors.json'), 'utf8'))

// The package does not export ./package.json, so the version is read from the
// installed tree rather than imported.
const version = JSON.parse(readFileSync(join(HERE, '..', '..', 'node_modules', 'agent-passport-system', 'package.json'), 'utf8')).version
const exported = Object.keys(aps).sort()

console.log(`agent-passport-system (npm) ${version}, ${exported.length} exports`)
console.log('')

// --- status_artifact: a real surface, run per vector ------------------------

console.log('group status_artifact: enforceFreshnessPolicy + isEvidenceFresh')
console.log('  mapping: AttestationFreshness {type: rotating, validAt: this_update,')
console.log('           ttl: next_update - this_update}, maxStalenessMs: the verifier bound.')
console.log('  The artifact\'s own declared refresh window becomes the evidence ttl and the')
console.log('  verifier\'s bound stays the verifier\'s bound, so both windows reach the SDK.')
console.log('')

for (const c of vectors.cases.filter(v => v.group === 'status_artifact')) {
  const a = c.input.artifact
  const ttlSeconds = (Date.parse(a.next_update) - Date.parse(a.this_update)) / 1000
  const freshness = aps.createRotatingFreshness(a.this_update, ttlSeconds)
  const now = new Date(c.input.evaluated_at)
  const opts = {
    source: a.list_id,
    maxStalenessMs: c.input.verifier_freshness_bound_s * 1000,
    checkedAt: now,
    freshness,
  }
  const own = aps.isEvidenceFresh(freshness, now)
  const closed = aps.enforceFreshnessPolicy(opts, { mode: 'fail_closed' })
  const bounded = aps.enforceFreshnessPolicy(opts, {
    mode: 'bounded_staleness',
    boundedStalenessMs: c.input.verifier_freshness_bound_s * 1000,
  })
  console.log(`  ${c.id}  fixture expects ${c.expected.verdict}`)
  console.log(`    isEvidenceFresh(own window)   : ${own}`)
  console.log(`    fail_closed                   : result=${closed.result} effect=${closed.effect} reason="${closed.reason}"`)
  console.log(`    bounded_staleness(bound=verifier): result=${bounded.result} effect=${bounded.effect} reason="${bounded.reason}"`)
}
console.log('')

// --- the other five groups: probe for a named surface -----------------------

const probes = [
  {
    group: 'status_correction',
    concept: 'retraction of an erroneous status publication, distinct from a revocation withdrawal',
    pattern: /(publicationError|retract|correctStatus|correctRevocation|unrevoke|withdrawRevocation)/i,
    reason:
      'no export names a publication-error retraction, and no export takes a list version range. The correction/withdrawal names the package does export (withdrawProvisional, withdrawalPayload) belong to the provisional-decision surface, not to a status publisher retracting its own published entry.',
  },
  {
    group: 'issuer_time_evidence',
    concept: 'independent time evidence covering an issuer-claimed issued_at',
    // Deliberately not matching a bare /timeAttestation/i: the export
    // `verifyRuntimeAttestation` contains that substring ("runTimeAttestation")
    // and is about runtime attestation, not about time. It is the only near-name
    // in the package and it is unrelated.
    pattern: /(timestampAuthority|trustedTimestamp|timeSourceDisagreement|rfc3161|externalTimeEvidence)/i,
    reason:
      'the package canonicalizes and asserts timestamps (assertCanonicalTimestamp, compareCanonicalTimestamps) but has no surface that takes evidence about a timestamp from a party other than the issuer, and none that takes a declared time-source-disagreement window.',
  },
  {
    group: 'history_reconciliation',
    concept: 'merging two divergent authority write histories with a reconciliation record',
    pattern: /(reconcileHistory|mergeHistories|splitBrain|partitionMerge|discardedWrites)/i,
    reason:
      'the revocation store is a single in-memory set (InMemoryAuthorityRevocationStore) with no notion of two histories, no merge entry point and no record of what a merge kept or discarded. reconcileBilateralPair is about a bilateral delegation pair, not about a partitioned authority store.',
  },
  {
    group: 'causal_read',
    concept: 'a required-observation token binding a read to a specific prior write',
    pattern: /(consistencyToken|zookie|zedToken|requiredObservation|readAfterWrite|appliedThrough)/i,
    reason:
      'the resolveRevocation callback the chain verifier accepts takes the delegation and nothing else, so there is no argument through which a request could name a write the answer must reflect.',
  },
  {
    group: 'evidence_coverage',
    concept: 'coverage of an evidence interval against a declared delivery lag and a consumer set',
    pattern: /(coverageOfInterval|evidenceCoverage|deliveryLag|deliveredThrough|intervalComplete)/i,
    reason:
      'capabilityCoverage is scope coverage, not interval coverage. Nothing takes a query interval, a declared delivery-lag bound, or a set of consumers with delivery positions.',
  },
]

for (const p of probes) {
  const hits = exported.filter(n => p.pattern.test(n))
  const verdict = hits.length === 0 ? 'not_supported' : `present: ${hits.join(', ')}`
  console.log(`group ${p.group}: ${verdict}`)
  console.log(`  needed: ${p.concept}`)
  console.log(`  reason: ${p.reason}`)
}

File: fixtures/lifecycle-infrastructure-failure/sdk_probe.py
#!/usr/bin/env python3
# Copyright 2026 Tymofii Pidlisnyi. Apache-2.0 license. See LICENSE.
#
# Records what the published PyPI `agent-passport-system` package offers for
# each group in this family, so SDK-RUNS.md's entries are reproducible rather
# than asserted.
#
# It reports, it does not judge. Nothing in draft-pidlisnyi-aps-03 asks the SDK
# for any of these six behaviours, so a `not_supported` line is an absence of a
# named surface, not a defect report and not a conformance result.
#
# The probe walks every importable module in the package and collects every
# public name, so an absence here is an absence across the whole package rather
# than across one module somebody remembered to look in.
#
# Run: /path/to/venv/bin/python fixtures/lifecycle-infrastructure-failure/sdk_probe.py
# Exit 0 always.

from __future__ import annotations

import importlib
import importlib.metadata
import pkgutil
import re
import sys

import agent_passport


def all_public_names() -> tuple[list[str], int]:
    modules: list[str] = []

    def walk(package, depth: int = 0) -> None:
        if depth > 4:
            return
        for info in pkgutil.iter_modules(package.__path__):
            full = f"{package.__name__}.{info.name}"
            modules.append(full)
            if info.ispkg:
                try:
                    walk(importlib.import_module(full), depth + 1)
                except Exception:
                    pass

    walk(agent_passport)
    names: set[str] = set()
    for module_name in modules:
        try:
            module = importlib.import_module(module_name)
        except Exception:
            continue
        for attribute in dir(module):
            if not attribute.startswith("_"):
                names.add(attribute)
    return sorted(names), len(modules)


PROBES = [
    (
        "status_artifact",
        "a freshness record carrying both a source's own declared validity window and the verifier's own bound",
        r"fresh|stale|max_staleness|next_update|evidence_age",
        "the Python package has no freshness, staleness or evidence-age surface at all. The npm package's "
        "enforceFreshnessPolicy / isEvidenceFresh / createRotatingFreshness have no Python counterpart at 4.1.0, so "
        "this group has no Python path and the npm run is the only SDK result this family has for it.",
    ),
    (
        "status_correction",
        "retraction of an erroneous status publication, distinct from a revocation withdrawal",
        r"retract|correct|unrevoke|withdraw|publication_error",
        "no name in the package retracts a published status entry, and none takes a list version range.",
    ),
    (
        "issuer_time_evidence",
        "independent time evidence covering an issuer-claimed issued_at",
        r"timestamp_authority|trusted_timestamp|time_source|rfc3161|external_time",
        "the package canonicalizes and compares timestamps (is_canonical_timestamp, compare_canonical_timestamps) "
        "but takes no evidence about a timestamp from a party other than the issuer, and no declared "
        "time-source-disagreement window.",
    ),
    (
        "history_reconciliation",
        "merging two divergent authority write histories with a reconciliation record",
        r"reconcile|merge|split_brain|partition|discarded",
        "InMemoryAuthorityRevocationStore is a single set with no notion of two histories, no merge entry point and "
        "no record of what a merge kept or discarded.",
    ),
    (
        "causal_read",
        "a required-observation token binding a read to a specific prior write",
        r"consistency_token|zookie|zed_token|required_observation|read_after_write|applied_through",
        "create_authority_revocation_resolver returns a callable over the delegation alone, so a request has no "
        "argument through which to name a write the answer must reflect.",
    ),
    (
        "evidence_coverage",
        "coverage of an evidence interval against a declared delivery lag and a consumer set",
        r"coverage|delivery_lag|delivered_through|interval_complete",
        "nothing takes a query interval, a declared delivery-lag bound, or a set of consumers with delivery "
        "positions.",
    ),
]


def main() -> int:
    version = importlib.metadata.version("agent-passport-system")
    names, module_count = all_public_names()
    print(f"agent-passport-system (PyPI) {version}, {len(names)} public names across {module_count} modules")
    print("")
    for group, concept, pattern, reason in PROBES:
        hits = [n for n in names if re.search(pattern, n, re.IGNORECASE)]
        verdict = "not_supported" if not hits else f"present: {', '.join(hits)}"
        print(f"group {group}: {verdict}")
        print(f"  needed: {concept}")
        print(f"  reason: {reason}")
    print("")
    print(
        "canonicalization: supported. agent_passport.canonical.canonicalize_jcs is what verify.py uses to recompute "
        "every case's pinned RFC 8785 input digest, so every vector in this family does have one PyPI SDK result."
    )
    return 0


if __name__ == "__main__":
    sys.exit(main())
